clippy_utils/src/ty/type_certainty/mod.rs

```diff
@@ -0,0 +1,310 @@
+//! A heuristic to tell whether an expression's type can be determined purely from its
+//! subexpressions, and the arguments and locals they use. Put another way, `expr_type_is_certain`
+//! tries to tell whether an expression's type can be determined without appeal to the surrounding
+//! context.
+//!
+//! This is, in some sense, a counterpart to `let_unit_value`'s `expr_needs_inferred_result`.
+//! Intuitively, that function determines whether an expression's type is needed for type inference,
+//! whereas `expr_type_is_certain` determines whether type inference is needed for an expression's
+//! type.
+//!
+//! As a heuristic, `expr_type_is_certain` may produce false negatives, but a false positive should
+//! be considered a bug.
+
+#![allow(clippy::wrong_self_convention)]
```

Comment:
I don't like bulk-allowing a lint like this, especially one like `wrong_self_convention`. We should probably put this on individual methods instead, as `expect` (PS: Which method is actually triggering it? I can't see any out of the ordinary.)

Reply:
> Which method is actually triggering it?

AFAICT, none are. (I can't remember why I included this originally.)